homeassistant/components/knx/config_flow.py

```diff
@@ -0,0 +1,265 @@
+        if self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
+
+        self._tunnels = []
+        return await self.async_step_type()
+
+    async def async_step_type(self, user_input: dict | None = None) -> FlowResult:
+        """Handle connection type configuration."""
+        errors: dict = {}
+        supported_connection_types = [CONF_KNX_TUNNELING, CONF_KNX_ROUTING]
+        fields = {}
+
+        if user_input is None:
+            gateways = await scan_for_gateways()
+
+            if gateways:
+                supported_connection_types.insert(0, CONF_KNX_AUTOMATIC)
+                self._tunnels = [
+                    gateway for gateway in gateways if gateway.supports_tunnelling
+                ]
+
+            fields = {
+                vol.Required(CONF_KNX_CONNECTION_TYPE): vol.In(
+                    supported_connection_types
+                )
+            }
+
+        if user_input is not None:
+            connection_type = user_input[CONF_KNX_CONNECTION_TYPE]
+            if connection_type == CONF_KNX_AUTOMATIC:
+                return self.async_create_entry(
+                    title=CONF_KNX_AUTOMATIC, data=user_input
+                )
+
+            if connection_type == CONF_KNX_ROUTING:
+                return await self.async_step_routing()
+
+            if connection_type == CONF_KNX_TUNNELING and self._tunnels:
+                return await self.async_step_tunnel()
+
+            return await self.async_step_manual_tunnel(user_input)
+
+        return self.async_show_form(
+            step_id="type", data_schema=vol.Schema(fields), errors=errors
+        )
+
+    async def async_step_manual_tunnel(
+        self, user_input: dict | None = None
+    ) -> FlowResult:
+        """General setup."""
+        errors: dict = {}
+
+        # prefill data for frontend
+        _gateway_ip: str = ""
+        _gateway_port: int = 3675
+        if user_input is not None and CONF_HOST in user_input:
+            _gateway_ip = user_input[CONF_HOST]
+            _gateway_port = user_input[CONF_PORT]
+
+        if user_input is not None and CONF_KNX_INDIVIDUAL_ADDRESS in user_input:
+            return self.async_create_entry(
+                title=user_input[CONF_HOST],
+                data={
+                    CONF_HOST: user_input[CONF_HOST],
+                    CONF_PORT: user_input[CONF_PORT],
+                    CONF_KNX_INDIVIDUAL_ADDRESS: user_input[
+                        CONF_KNX_INDIVIDUAL_ADDRESS
+                    ],
+                    ConnectionSchema.CONF_KNX_ROUTE_BACK: user_input[
+                        ConnectionSchema.CONF_KNX_ROUTE_BACK
+                    ],
+                    CONF_KNX_CONNECTION_TYPE: CONF_KNX_TUNNELING,
+                },
+            )
+
+        fields = {
+            vol.Required(CONF_HOST, default=_gateway_ip): str,
+            vol.Required(CONF_PORT, default=_gateway_port): vol.Coerce(int),
```

Comment:
This adds a min/max too:
```suggestion
            vol.Required(CONF_PORT, default=_gateway_port): cv.port,
```